Implement --delete-generations + flag for keeping last N number of gens

mogorman opened the pull request:
I don't find keeping last X days of generations as useful to me as keeping the last X number of generations.  so I implemented it.

domenkozar commented:
cc @edolstra

edolstra commented on src/libstore/profiles.hh:
```diff
@@ -37,6 +37,8 @@ void deleteGeneration(const Path & profile, unsigned int gen);
 
 void deleteGenerations(const Path & profile, const std::set<unsigned int> & gensToDelete, bool dryRun);
 
+void deleteGenerationsGreaterThan(const Path & profile, const string & max, bool dryRun);
```
`max` should be an integer. (I.e. the conversion to int should be done in nix-env.cc where the argument parsing happens.)

edolstra commented on src/libstore/profiles.cc:
```diff
@@ -157,6 +157,29 @@ void deleteGenerations(const Path & profile, const std::set<unsigned int> & gens
     }
 }
 
+void deleteGenerationsGreaterThan(const Path & profile, const string & max, bool dryRun)
```
Indentation is a bit off in this function (and there are a few tabs).

edolstra commented:
Looks good apart from my two comments.

mogorman commented:
fixes what was brought up in comments

Profpatsch commented:
merge?

mogorman commented:
bump from mailinglist

mogorman commented:
@edolstra  or @domenkozar could you take a look if you get the time?

jb55 commented on src/nix-env/nix-env.cc:
```diff
@@ -1285,6 +1285,14 @@ static void opDeleteGenerations(Globals & globals, Strings opFlags, Strings opAr
+    } else if (opArgs.size() == 1 && opArgs.front().find('+') != string::npos) {
+        if(opArgs.front().size() < 2)
+            throw Error(format("invalid number of generations ‘%1%’") % opArgs.front());
+        string str_max = string(opArgs.front() 1, opArgs.front().size());
```
typo here? `opArgs.front() 1`

mogorman commented:
yup you were right i hadnt rerun it when i made that change

edolstra commented on src/libstore/profiles.hh:
```diff
@@ -37,6 +37,8 @@ void deleteGeneration(const Path & profile, unsigned int gen);
 
 void deleteGenerations(const Path & profile, const std::set<unsigned int> & gensToDelete, bool dryRun);
 
+void deleteGenerationsGreaterThan(const Path & profile, const int max, bool dryRun);
```
Can `deleteGenerations` be removed? It looks like a special case of `deleteGenerationsGreaterThan` with `max=1`.

mogorman replied:
No, ```deleteGenerations``` can be given a list of specific generations where as mine will only keep x in chronological order, or the date one will only keep so many dates back.

mogorman replied:
@edolstra any chance we can get this merged before 18.03 release?

copumpkin replied:
I suppose we could unify the two functions by simply having `deleteGenerationsOlderThan` build the `std::set` and then pass it into `deleteGenerations`. I doubt the deduplication would be that valuable though.

Although I did notice that you don't check against `curGen`, but `deleteGenerations` does. Do you do anything to ensure you don't delete the current generation? If not, perhaps it'd pay to just call the existing function with the set you computed.

mogorman replied:
@copumpkin I could be mistaken but rereading the logic I wrote, it shouldn't be possible to delete the current generation, the list of gens checks to see if it hits the end of generations and bails out at that point.

mogorman replied:
@copumpkin you were correct that it would have been possible to have deleted the currentgeneration if you had done enough rollbacks. I have modified the code to make sure this is not possible

dtzWill commented:
Is the last generation always the current one, even after rollback and such?

mogorman commented:
I am going to confirm the delete current generation issue now in a roll back situation now.

mogorman commented:
thanks @dtzWill i went back and changed the logic slightly. now it starts the count from the current generation rather than the beginning of the list. I also updated docs to explain this. I think this is more likely what people will want. I would say this is good to go again.

jackjennings commented:
Any chance this is still on the table? This would help put immensely in a nixops managed server that consistently runs out of space unless cleaned up manually. Being able to only keep the last N generations would make keeping the disk relatively empty much easier.

I would be happy to contribute if some manual tasks need to be performed to update this pull request to work with the latest release…

peti commented:
This will do it:

~~~~
$ cd /nix/var/nix/profiles
$ ls -1d system-*-link | head -n -2 | xargs --no-run-if-empty rm
# nix-collect-garbage
~~~~

mogorman commented:
woohoo

jackjennings commented:
Thank you @peti!
